casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -65,191 +66,414 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
+   Two incoming sends/receives, at most one had a matching dual in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They couldn't be competing for the same linear receive/send (at most one had a match).
+   Notice this includes "two unsatisfied" and "must be looking for different data" cases.
+    */
+  object SamePolarityMerge extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   Two incoming sends/receives each matched a receive/send that was in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They could've matched the same linear event.
+   Mergeable if different events, or at least one matched event is non-linear.
+   This is the case where both incoming events could have matched what was in the other TS.
+    */
+  object CouldMatchSameConflicts extends ConflictingCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(allMatch(left ++ right, base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   Two incoming sends/receives each matched a receive/send that was in TS.
+   The incoming events won't cause more COMMs together (same polarity).
+   They could've matched the same linear event.
+   Mergeable if different events, or at least one matched event is non-linear.
+   This is the case where the incoming events match differently
+    */
+  object CouldMatchSameMerges extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      samePolarities(left)(right)
+      differentPolarities(left)(base)
+      differentPolarities(right)(base)
+      assert(!allMatch(left ++ right, base) || onePersistent(base))
+    }.pure[Effect]
+
+  }
+
+  /***
+   A send and a receive were incoming, at least one had a match, either:
+   - both were linear
+   - one was non-linear, the other had a match
+   They couldn't match the same linear event (they have different polarity)
+   They couldn't spawn more work, because either:
+   - both were linear, one of them had a match in TS
+   - one was non-linear, but the other chose to go with its match
+    */
+  object HadItsMatch extends MergeableCase {
+
+    override def precondition(
+        left: Rho*
+    )(right: Rho*)(base: Rho*)(implicit pos: source.Position): Effect[_] = {
+      differentPolarities(left)(right)
+      crossedPolarity(left)(base)
+      crossedPolarity(right)(base)
+      assert(someMatch(left, base) || someMatch(right, base))
```

Comment:
it'd be great to encode the
```
 either:
   - both were linear
   - one was non-linear, the other had a match
```
criteria for the incoming events.

With singular left/right:
```
val (a, b) = if (left.cardinality == Linear) (left, right) else (right, a)
assert(b.cardinality == Linear || hasMatch(a, base))
```